x/tx/aminojson/encoder.go

```diff
@@ -0,0 +1,171 @@
+			return err
+		}
+		return aj.marshalList(list, w)
+	default:
+		return fmt.Errorf("unsupported type %T", list)
+	}
+}
+
+// keyFieldEncoder replicates the behavior at described at:
+// https://github.com/cosmos/cosmos-sdk/blob/b49f948b36bc991db5be431607b475633aed697e/proto/cosmos/crypto/secp256k1/keys.proto#L16
+// The message is treated if it were bytes directly without the key field specified.
+func keyFieldEncoder(msg protoreflect.Message, w io.Writer) error {
+	keyField := msg.Descriptor().Fields().ByName("key")
+	if keyField == nil {
+		return errors.New(`message encoder for key_field: no field named "key" found`)
+	}
+
+	bz := msg.Get(keyField).Bytes()
+
+	if len(bz) == 0 {
+		_, err := fmt.Fprint(w, "null")
+		return err
+	}
+
+	_, err := fmt.Fprintf(w, `"%s"`, base64.StdEncoding.EncodeToString(bz))
+	return err
+
+}
+
+type moduleAccountPretty struct {
+	Address       string   `json:"address"`
+	PubKey        string   `json:"public_key"`
+	AccountNumber uint64   `json:"account_number"`
+	Sequence      uint64   `json:"sequence"`
+	Name          string   `json:"name"`
+	Permissions   []string `json:"permissions"`
+}
+
+// moduleAccountEncoder replicates the behavior in
+// https://github.com/cosmos/cosmos-sdk/blob/41a3dfeced2953beba3a7d11ec798d17ee19f506/x/auth/types/account.go#L230-L254
+func moduleAccountEncoder(msg protoreflect.Message, w io.Writer) error {
+	ma := msg.Interface().(*authapi.ModuleAccount)
+	pretty := moduleAccountPretty{
+		PubKey:      "",
+		Name:        ma.Name,
+		Permissions: ma.Permissions,
+	}
+	if ma.BaseAccount != nil {
+		pretty.Address = ma.BaseAccount.Address
+		pretty.AccountNumber = ma.BaseAccount.AccountNumber
+		pretty.Sequence = ma.BaseAccount.Sequence
+	} else {
+		pretty.Address = ""
+		pretty.AccountNumber = 0
+		pretty.Sequence = 0
+	}
+
+	bz, err := json.Marshal(pretty)
+	if err != nil {
+		return err
+	}
+	_, err = w.Write(bz)
+	return err
+}
+
+// thresholdStringEncoder replicates the behavior at:
+// https://github.com/cosmos/cosmos-sdk/blob/4a6a1e3cb8de459891cb0495052589673d14ef51/crypto/keys/multisig/amino.go#L35
+// also see:
+// https://github.com/cosmos/cosmos-sdk/blob/b49f948b36bc991db5be431607b475633aed697e/proto/cosmos/crypto/multisig/keys.proto#L15/
+func thresholdStringEncoder(msg protoreflect.Message, w io.Writer) error {
+	pk, ok := msg.Interface().(*multisig.LegacyAminoPubKey)
+	if !ok {
+		return errors.New("thresholdStringEncoder: msg not a multisig.LegacyAminoPubKey")
+	}
+	_, err := w.Write([]byte(fmt.Sprintf(`{"threshold":"%d","pubkeys":`, pk.Threshold)))
+	if err != nil {
+		return err
+	}
+
+	if len(pk.PublicKeys) == 0 {
+		_, err = w.Write([]byte(`[]}`))
+		return err
+	}
+
+	fields := msg.Descriptor().Fields()
+	pubkeysField := fields.ByName("public_keys")
+	pubkeys := msg.Get(pubkeysField).List()
+
+	aj := Encoder{}.DefineMessageEncoding("key_field", keyFieldEncoder)
+	err = aj.marshalList(pubkeys, w)
+	if err != nil {
+		return err
+	}
```

Comment:
Re-looking at this PR, i'm wondering if it's worth changing the message encoder func signature to:

```diff
- type MessageEncoder func(protoreflect.Message, io.Writer) error
+ type MessageEncoder func(Encoder, protoreflect.Message, io.Writer) error
```

this way we don't need to create a new instance of `Encoder{}` here.

Reply:
It should be *Encoder in that case I think

Reply:
> It should be *Encoder in that case I think

Why do you say so?

Reply:
Efficiency